xrmocap/core/evaluation/metrics.py

```diff
@@ -0,0 +1,200 @@
+    # There must be no np.nan in the pred_keypoints3d
+    mpjpe, pa_mpjpe, pck_50, pck_100 = [], [], [], []
+    n_frame = gt_keypoints3d.get_frame_number()
+    gt_kps3d = gt_keypoints3d.get_keypoints()[..., :3]
+    gt_kps3d_mask = gt_keypoints3d.get_mask()
+    pred_kps3d = pred_keypoints3d.get_keypoints()[..., :3]
+    pred_kps3d_mask = pred_keypoints3d.get_mask()
+    pred_kps3d_convention = pred_keypoints3d.get_convention()
+    gt_kps3d_convention = gt_keypoints3d.get_convention()
+    for frame_idx in range(n_frame):
+        if not gt_kps3d_mask[frame_idx].any():
+            continue
+        gt_kps3d_idxs = np.where(
+            np.sum(gt_kps3d_mask[frame_idx], axis=1) > 0)[0]
+        for gt_kps3d_idx in gt_kps3d_idxs:
+            f_gt_kps3d = gt_kps3d[frame_idx][gt_kps3d_idx]
+            f_pred_kps3d = pred_kps3d[frame_idx][
+                np.sum(pred_kps3d_mask[frame_idx], axis=1) > 0]
+            if len(f_pred_kps3d) == 0:
+                continue
+
+            dist = vectorize_distance(f_gt_kps3d[np.newaxis], f_pred_kps3d)
+            f_pred_kps3d = f_pred_kps3d[np.argmin(dist[0])]
+
+            if np.all((f_pred_kps3d == 0)):
+                continue
+
+            # MPJPE
+            f_pred_keypoints = Keypoints(
+                kps=np.concatenate(
+                    (f_pred_kps3d, np.ones_like(f_pred_kps3d[..., 0:1])),
+                    axis=-1),
+                convention=pred_kps3d_convention)
+            f_gt_keypoints = Keypoints(
+                kps=np.concatenate(
+                    (f_gt_kps3d, np.ones_like(f_gt_kps3d[..., 0:1])), axis=-1),
+                convention=gt_kps3d_convention)
+            mpjpe.append(
+                compute_mpjpe(f_pred_keypoints, f_gt_keypoints, align=True))
+
+            # PA-MPJPE
+            _, _, rotation, scaling, transl = compute_similarity_transform(
+                f_gt_kps3d, f_pred_kps3d, compute_optimal_scale=True)
+            pred_kps3d_pa = (scaling * f_pred_kps3d.dot(rotation)) + transl
+
+            pred_keypoints_pa = Keypoints(
+                kps=np.concatenate(
+                    (pred_kps3d_pa, np.ones_like(pred_kps3d_pa[..., 0:1])),
+                    axis=-1),
+                convention=pred_kps3d_convention)
+            pa_mpjpe_i = compute_mpjpe(
+                pred_keypoints_pa, f_gt_keypoints, align=True)
+            pa_mpjpe.append(pa_mpjpe_i)
+
+            pck_50.append(np.mean(pa_mpjpe_i <= (pck_50_thres / scale)))
+            pck_100.append(np.mean(pa_mpjpe_i <= (pck_100_thres / scale)))
+    mpjpe = np.asarray(mpjpe) * scale  # m to mm
+    pa_mpjpe = np.asarray(pa_mpjpe) * scale  # m to mm
+    mpjpe_mean, mpjpe_std = np.mean(mpjpe), np.std(mpjpe)
+    pa_mpjpe_mean, pa_mpjpe_std = np.mean(pa_mpjpe), np.std(pa_mpjpe)
+    pck_50 = np.mean(pck_50) * 100.  # percentage
+    pck_100 = np.mean(pck_100) * 100.  # percentage
+    logger.info(f'MPJPE: {mpjpe_mean:.2f} ± {mpjpe_std:.2f} mm')
```

Comment:
`def evaluate()` soulds like a function for calculating, not logging. We may return `pck_50, pck_100, mpjpe` or a dict like  `dict(pck_50=pck_50, pck_100=pck_100...)`, and let the caller decide whether to log it or keep the console silent.